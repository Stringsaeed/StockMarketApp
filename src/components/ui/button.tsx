import {ComponentProps} from 'react';
import {Gesture, GestureDetector} from 'react-native-gesture-handler';
import Animated, {
  interpolate,
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from 'react-native-reanimated';

interface ButtonProps extends ComponentProps<typeof Animated.View> {
  children: React.ReactNode;
}

export const Button = ({style: userStyle, children, ...props}: ButtonProps) => {
  const pressProgress = useSharedValue(0);

  const tap = Gesture.Tap()
    .onEnd(() => {
      pressProgress.value = withSpring(1);
    })
    .onFinalize(() => {
      pressProgress.value = withSpring(0);
    })
    .onStart(() => {
      pressProgress.value = withSpring(0);
    })
    .onBegin(() => {
      pressProgress.value = withSpring(0.5);
    });

  const style = useAnimatedStyle(() => {
    return {
      transform: [
        {scale: interpolate(pressProgress.value, [0, 1], [1, -0.95])},
      ],
    };
  });

  return (
    <GestureDetector gesture={tap}>
      <Animated.View {...props} style={[userStyle, style]}>
        {children}
      </Animated.View>
    </GestureDetector>
  );
};
