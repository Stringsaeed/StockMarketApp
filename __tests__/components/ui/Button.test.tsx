import {fireEvent, render, screen} from '@testing-library/react-native';
import React from 'react';
import {Text, View} from 'react-native';

import {Button} from '@/components/ui/button';

// Mock the react-native-gesture-handler
jest.mock('react-native-gesture-handler', () => {
  return {
    Gesture: {
      Tap: () => ({
        onFinalize: jest.fn().mockReturnThis(),
        onBegin: jest.fn().mockReturnThis(),
        onStart: jest.fn().mockReturnThis(),
        onEnd: jest.fn().mockReturnThis(),
      }),
    },
    GestureDetector: (props: any) => props.children,
  };
});

describe('Button Component', () => {
  it('renders correctly', () => {
    render(
      <Button testID="test-button">
        <Text>Test Button</Text>
      </Button>,
    );

    expect(screen.getByTestId('test-button')).toBeTruthy();
    expect(screen.getByText('Test Button')).toBeTruthy();
  });

  it('matches snapshot', () => {
    render(
      <Button>
        <Text>Test Button</Text>
      </Button>,
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('applies custom styles', () => {
    const customStyle = {backgroundColor: 'red', borderRadius: 8};

    render(
      <Button style={customStyle} testID="styled-button">
        <Text>Styled Button</Text>
      </Button>,
    );

    const button = screen.getByTestId('styled-button');
    expect(button).toBeTruthy();
  });

  it('passes additional props', () => {
    render(
      <Button testID="props-button" accessibilityLabel="Accessible Button">
        <Text>Props Button</Text>
      </Button>,
    );

    const button = screen.getByTestId('props-button');
    expect(button).toBeTruthy();
  });

  it('renders children correctly', () => {
    render(
      <Button testID="children-button">
        <Text testID="button-text">Button Text</Text>
        <View testID="button-icon" />
      </Button>,
    );

    expect(screen.getByTestId('children-button')).toBeTruthy();
    expect(screen.getByTestId('button-text')).toBeTruthy();
    expect(screen.getByTestId('button-icon')).toBeTruthy();
  });

  it('handles press animations', () => {
    render(
      <Button testID="animated-button">
        <Text>Animated Button</Text>
      </Button>,
    );

    const button = screen.getByTestId('animated-button');

    // Simulate press
    fireEvent(button, 'onTouchStart');
    fireEvent(button, 'onTouchEnd');

    expect(button).toBeTruthy();
  });
});
