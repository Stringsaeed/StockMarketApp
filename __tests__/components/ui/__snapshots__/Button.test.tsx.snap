// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches snapshot 1`] = `
<View
  style={
    [
      undefined,
      {
        "transform": [
          {
            "scale": undefined,
          },
        ],
      },
    ]
  }
>
  <Text>
    Test Button
  </Text>
</View>
`;
